### PR TITLE
Add tensor algorithm for all pairs CFL-reachability problem

### PR DESCRIPTION
**Pull request by vkutuev**
- Add RSM decomposed into boolean matrices (BooleanMatrixRsm)
- Add tensor algorithm for all pairs CFL-reachability problem
- Update requirements versions: pyformlang and cfpq-data
- Add tests for tensor CFL-reachability algorithm
- Update README

### CODE REVIEW

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
+"""Function that solves the All-Pairs CFL-reachability problem
+using boolean matrix Kronecker product"""
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: using the Kronecker product of Boolean matrices

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+from typing import Tuple, List
+
+import networkx as nx
+
+from pyformlang.cfg import CFG
+from pyformlang.rsa import RecursiveAutomaton
+from pygraphblas import Matrix, BOOL
+
+from cfpq_pyalgo.pygraphblas import (
+    BooleanMatrixGraph,
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: These class names is like we have one Boolean matrix for a graph. What about names "GraphBooleanDecomposition" and "RSMBooleanDecomposition"? Could these classes share some common "BooleanDecomposition" interface or should we leave them separated?

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    BooleanMatrixRsm,
+    bmg_from_nx_graph,
+)
+
+
+def tensor_all_pairs_reachability(
+    graph: nx.MultiDiGraph, grammar: CFG
+) -> List[Tuple[int, int]]:
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: The result for the reachability algorithms is a set of vertex pairs. Should we use Set[Tuple[int, int]]? When we compare results of different reachability algorithms, we should compare the set of vertex pairs, not lists.

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    """Determines the pairs of vertices (u, v) where there exists a path from u to v
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: Determines the set of vertex pairs (u, v) in `graph` such that there is a path from u to v

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    in `graph` and its word is in the language of Context-Free Grammar `grammar`.
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: whose edge labels form a word from the language generated by the context-free grammar `grammar`.

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    graph: `nx.MultiDiGraph`
+        NetworkX MultiDiGraph
+
+    grammar: `CFG`
+        Context-free grammar
+
+    Returns
+    -------
+    pairs: `List[Tuple[int, int]]`
+        List of CFL-reachable vertices pairs
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: Set of reachable vertex pairs

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py (hunk 3 above)
Comment (rustam-azimov) [reply cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: Can we support any type of graph nodes, not only int?

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+) -> Tuple[BooleanMatrixGraph, Matrix]:
+    graph_size = graph.matrices_size
+    # Initialize matrices for all labels
+    t = BooleanMatrixGraph(graph.matrices_size)
+    for label in rsm.labels:
+        if label in graph:
+            t[label] = graph[label].dup()
+        else:
+            t[label] = Matrix.sparse(BOOL, graph_size, graph_size)
+    # 0. Boxes with epsilon path
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py:75]: Please fix enumeration of steps in the comments

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+                ]
+            control_sum = t[nonterminal].nvals
+            t[nonterminal] += block
+            new_control_sum = t[nonterminal].nvals
+            if new_control_sum != control_sum:
+                changed = True
+    return t, kron
+
+
+def _transitive_closure(m: Matrix):
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py:111]: Please, remind me again why we are calculate degrees step by step? Why here we don't use the repeated squaring optimization (more dense matrices but log number of iterations)? Like m += m @ m. Please check if it faster on our real-world graphs.

File: cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+from typing import Dict, List, Set, Tuple
+
+from pyformlang.finite_automaton import State, Symbol
+from pyformlang.rsa import Box, RecursiveAutomaton
+from pygraphblas import Matrix
+
+from cfpq_pyalgo.pygraphblas import BooleanMatrixGraph
+
+
+class BooleanMatrixRsm:
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py]: Rsm -> RSM

File: cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        """All RSM labels"""
+        return self._labels
+
+    @property
+    def nonterminals(self) -> Set[str]:
+        """Labels of boxes"""
+        return self._nonterminals
+
+    def add_edge(self, u: int, v: int, label: str) -> None:
+        """Add an edge between `u` and `v` states with label `label.
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py]: with label `label`.

File: cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+
+        The states `u` and `v` will be automatically added if they are
+        not already in the RSM.
+
+        Parameters
+        ----------
+        u: int
+            The tail of the edge
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py]: tail and head need to be swapped? Or u is the tail of edge?

File: cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+        v: int
+            The head of the edge
+
+        label: str
+            The label of the edge
+        """
+        self._transitions.add_edge(u, v, label)
+
+    def get_start_state(self, label: str) -> int:
+        """Get start vertex for RSM box with label `label`
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py]: Some doc strings with dot at the end, and some strings without dots. Please add dots everywhere.

File: cfpq_pyalgo/pygraphblas/grammars/wcnf.py [hunk 13]
```diff
@@ -79,7 +79,7 @@ def contains(self, word: str):
         return self._cfg.contains(word)
 
     @classmethod
-    def from_text(cls, text: str, start_variable: Variable = Variable("S")):
+    def from_text(cls, text: str, start_variable: Variable = Variable("S")) -> "WCNF":
         cfg = CFG.from_text(text, start_variable)
         return cls(cfg)
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/grammars/wcnf.py:84]: Why we want to return "WCNF" here?

**Review by rustam-azimov — CHANGES_REQUESTED** [review]

File: cfpq_pyalgo/pygraphblas/grammars/wcnf.py (hunk 13 above)
Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/grammars/wcnf.py:84]: This is the "second constructor". This function creates an instance of WCNF from the grammar represented as text and returns it. So return type must be `WCNF` (but double quotes are required to use the class name in its definition)

File: cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py (hunk 11 above)
Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py]: *tail* is the source vertex and *head* is the destination vertex

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py (hunk 5 above)
Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: Fixed

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py (hunk 4 above)
Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: Fixed

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py (hunk 3 above)
Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: Return type changed to `Set[Tuple[Hashable, Hashable]]` (vertices of `nx.MultiDiGraph` are `Hashable` objects)

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py (hunk 2 above)
Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: Class names changed to `GraphBooleanDecomposition` and `RSMBooleanDecomposition`

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py (hunk 6 above)
Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: Fixed

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py (hunk 7 above)
Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py:75]: Fixed

File: cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py (hunk 9 above)
Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py]: Fixed

File: cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py (hunk 12 above)
Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py]: Fixed

File: cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py (hunk 1 above)
Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/algorithms/tensor_all_pairs_reachability.py]: Fixed

File: cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py (hunk 10 above)
Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/grammars/boolean_matrix_rsm.py]: Fixed

File: cfpq_pyalgo/pygraphblas/algorithms/matrix_all_pairs_reachability.py [hunk 14]
```diff
@@ -1,14 +1,14 @@
 """Function that solves the All-Pairs CFL-reachability problem
-using boolean matrix multiplication"""
+using boolean matrix multiplication."""
```
Comment (rustam-azimov) [inline cfpq_pyalgo/pygraphblas/algorithms/matrix_all_pairs_reachability.py]: using Boolean matrix multiplication.

Comment (vkutuev) [reply cfpq_pyalgo/pygraphblas/algorithms/matrix_all_pairs_reachability.py]: Fixed

**Review by rustam-azimov — APPROVED** [review]

**Pull request merged**